libr/anal/p/anal_avr.c

```diff
@@ -17,35 +17,277 @@
+	CPU_MODEL_DECL ("ATmega168",  13, 512),
+	CPU_MODEL_DECL ("ATmega640",  16, 512),
+	CPU_MODEL_DECL ("ATmega1280", 16, 512),
+	CPU_MODEL_DECL ("ATmega1281", 16, 512),
+	CPU_MODEL_DECL ("ATmega2560", 22, 512),
+	CPU_MODEL_DECL ("ATmega2561", 22, 512),
+	CPU_MODEL_DECL ((char *) 0,   16, 512) };
+
+INST_HANDLER (cp) {
+	// CP Rd, Rr
+	int r = (buf[0]        & 0x0f) | ((buf[1] << 3) & 0x10);
+	int d = ((buf[0] >> 4) & 0x0f) | ((buf[1] << 4) & 0x10);
+
+	r_strbuf_setf (
+		&op->esil,
+		"r%d,r%d,==,"			// compare Rr vs Rd
+		"$z,zf,=,"			// zero flag (zf)
+		"r%d,r%d,^,0x08,!,!,hf,"	// half carry flag (hf)
+		"r%d,r%d,^,0x80,!,!,cf,"	// carry flag (cf)
+		"$o,vf,=,"			// overflow flag (vf)
+		"r%d,r%d,-,0x80,&,!,!,nf,=,"	// neg flg: result's MSB is set
+		"vf,nf,^,sf,=",			// sign flag: xor(vf, nf)
+		r, d, r, d, r, d, r, d);
+		//"r%d,r%d,==,"			// compare Rr vs Rd
+		//"$z,zf,=,"			// zero flag (zf)
+		//"$b3,hf,=,"			// half carry flag (hf)
+		//"$b8,cf,=,"			// carry flag (cf)
+		//"$o,vf,=,"			// overflow flag (vf)
+		//"r%d,r%d,-,0x80,&,!,!,nf,=,"	// neg flg: result's MSB is set
+		//"vf,nf,^,sf,=",			// sign flag: xor(vf, nf)
+}
+
+INST_HANDLER (cpc) {
+	int r = (buf[0]        & 0x0f) | ((buf[1] << 3) & 0x10);
+	int d = ((buf[0] >> 4) & 0x0f) | ((buf[1] << 4) & 0x10);
+
+	r_strbuf_setf (
+		&op->esil,
+		"r%d,cf,r%d,-,0xff,&,-,"	// ((Rd - C) & 0xff) - Rr
+		"0x80,&,!,!,nf,=,"		// nef flg: result's MSB is set
+		"r%d,cf,r%d,-,0xff,&,==,"	// ((Rd - C) & 0xff) == Rr
+		"$z,zf,=,"			// zero flag
+		"$b8,cf,=,"			// carry flag
+		"$b3,hf,=,"			// half carry flag
+		"$o,vf,=,"			// overflow flag
+		"vf,nf,^,sf,=",			// sign flag : xor(vf, nf)
+		r, d, r, d);
+}
+
+// ALIAS: Opcode (clr) when Rra and Rrb are the same
+INST_HANDLER (eor) {
+	int d = ((buf[0] & 0xf0) >> 4) | ((buf[1] & 1) << 4);
+	int r = (buf[0] & 0xf) | ((buf[1] & 2) << 3);
+
+	r_strbuf_setf (&op->esil, "r%d,r%d,^=,$z,zf,=,r%d,0x80,&,!,!,nf,=,nf,sf,=,0,vf,=", r, d, d);
+}
+
+INST_HANDLER (movw) {
+	int d = (buf[0] & 0xf0) >> 3;
+	int r = (buf[0] & 0x0f) << 1;
+
+	r_strbuf_setf (&op->esil, "r%d,r%d,=,r%d,r%d,=", r, d, r + 1, d + 1);
+}
+
+INST_HANDLER (nop) {
+}
+
+INST_HANDLER (out) {
+	op->type2 = 1;
+	op->val = (buf[0] & 0x0f) | (((buf[1] >> 1) & 0x03) << 4);
+	
+	int r = ((buf[0] >> 4) & 0x0f) | ((buf[1] & 0x01) << 4);
+
+	// launch esil trap (communicate upper layers about this I/O)
+	switch (op->val) {
+	case 0x3f: /* SREG */ r_strbuf_setf (&op->esil, "r%d,sreg,=", r); break;
+	case 0x3e: /* SPH  */ r_strbuf_setf (&op->esil, "r%d,sph,=",  r); break;
+	case 0x3d: /* SPL  */ r_strbuf_setf (&op->esil, "r%d,spl,=",  r); break;
+	default:
+		r_strbuf_setf (&op->esil, "2,$");
+	}
+}
+
+INST_HANDLER (rcall) {
+	register int32_t offset;
+
+	// target offset
+	offset = ((((buf[1] & 0xf) << 8) | buf[0]) + 1) << 1;
+	if (offset & 0x100)
+		offset |= 0xfffff000;
+	op->jump = op->addr + offset;
+	op->fail = 0;
+
+	// cycles!
+	if (strncasecmp (anal->cpu, "ATtiny", 6)) {
```

Comment:
iwould do the "true" block first, so prefix that condition with ! to make clear that